### ISSUE TITLE
Switch network doesn't work with non default chains

### ISSUE THREAD

**Issue by layinka** [tmm/wagmi#180]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

0.2.7

### Current Behavior

I added some new chains when setting up the providers.

When i use SwitchNetwork from the useNetwork  hook with chains i added e.g BSC, BSC Testnet etc, the selected network and list of networks do not show any longer.
There is no error and if i refresh the page, it now shows that i am connected to the right network

### Expected Behavior

SwitchNetwork should work well with default and non default chains

### Steps To Reproduce

Add new chains to the provider on setup and then use the code below to select and switch between chains

```
import { useNetwork } from 'wagmi'

const App = () => {
  const [{ data, error, loading }, switchNetwork] = useNetwork()

  return (
    <>
      <div>
        {data.chain?.name ?? networkData.chain?.id}{' '}
        {data.chain?.unsupported && '(unsupported)'}
      </div>

      {switchNetwork &&
        data.chains.map((x) =>
          x.id === data.chain?.id ? null : (
            <button key={x.id} onClick={() => switchNetwork(x.id)}>
              Switch to {x.name}
            </button>
          ),
        )}

      {error && <div>{error?.message}</div>}
    </>
  )
}
```

**Comment by o-az** [comment]
I'm using a workaround for this using utilizing `window.ethereum`. This solution doesn't rely on any library as long as you're targeting browser MetaMask users. You can modify my solution to support your use case:

```ts
// We'll use this in a few places
const ETHEREUM_CHAIN_ID = '0x1';
```
First, I have an `addEthereumChain` helper method that I use in the (unlikely) case that user does not have Ethereum chain added to their wallet, it will trigger user to add ethereum chain. We'll use this helper in the next part:
```ts
const addEthereumChain = async () => {
  if (typeof window.ethereum === 'undefined') return;
  return await window.ethereum.request({
    method: 'wallet_addEthereumChain',
    params: [
      {
        chainId: ETHEREUM_CHAIN_ID,
        chainName: 'Ethereum',
        rpcUrls: [`https://mainnet.infura.io/v3/${INFURA_ID}`],
      },
    ],
  });
};
```

Next I have the `switchNetwork` function that allows me to trigger MetaMask to ask user to switch to Ethereum network:
```ts
const switchNetwork = async () => {
  if (typeof window.ethereum === 'undefined') return;
  try {
    await window?.ethereum?.request({
      method: 'wallet_switchEthereumChain',
      params: [{ chainId: ETHEREUM_CHAIN_ID }],
    });
    return true;
  } catch (error) {
    if ((error as any).code !== 4902) return;
    await addEthereumChain();
  }
  return false;
};
```
Then I have a `switchOnChainChange` to switch to Eth network when chain changes while user is on:
```ts
const switchOnChainChange = async () => {
  if (typeof window.ethereum?.on === 'undefined') return;

  return window.ethereum.on('chainChanged', (chainId: string) => {
    console.log(`chain changed to ${chainId}`);

    if (chainId !== ETHEREUM_CHAIN_ID) {
      switchNetwork();
      // HERE: do whatever else you want, I'm triggering a toast
    } else {
      // HERE: do whatever else you want, I'm removing the toast on valid chain switch
    }
  });
};

```

Now that we have all of this, here's the `React` custom hook that will do the rest of the work:
```ts
export const useSwitchToEthereum = () => {
  return React.useEffect(() => {
    if (typeof window === 'undefined' || typeof window.ethereum === 'undefined') return;

    // prompts user to switch to Eth network when chain changes while user is on
    switchOnChainChange();

    // This is for when user first lands on the page and is not on the ethereum chain
    // @ts-ignore
    if (window.ethereum.chainId !== ETHEREUM_CHAIN_ID) {
      switchNetwork();
      // HERE: do whatever else you want, I'm triggering a toast that loads on page load
    } else {
      // HERE: do whatever else you want, I'm removing the toast on valid chain switch
    }
  }, []);
};
```
Finally, import this custom hook and add it in your top component (I'm using `next.js` so I have it in my `_app.tsx` file):

```ts
const App = ({ Component, pageProps }: AppProps) => {

  useSwitchToEthereum();

  return <Component {...pageProps} />;
};

export default App;
```

I have yet to thoroughly test this so I don't recommend adding this to production code until you've done your testing. So far, it's been working great for me.

If you have any issues feel free to reach out on twitter [@undeterrable](https://twitter.com/undeterrable)

**Comment by Stono** [comment]
+1 to this, was about to raise the same issue haha!   Thanks so much for sharing your work around

**Comment by tmm** [comment]
Looks like there's an issue with MetaMask, where is emits a `disconnect` event before switching chain.
https://github.com/MetaMask/metamask-extension/issues/13375#issuecomment-1046125113

Until this gets fixed, I'm adding a shim to help prevent this from happening. https://github.com/tmm/wagmi/pull/190

**Comment by layinka** [comment]
OK thanksThough, the bug doesn't happen with default chains in wagmi-------- Original message --------From: awkweb ***@***.***>Date: Sun, 20 Feb 2022, 01:00To: tmm/wagmi ***@***.***>Cc: layinka ***@***.***>, Author ***@***.***>Subject: Re: [tmm/wagmi] Switch network doesn't work with non default chains (Issue #180)
Looks like there's an issue with MetaMask, where is emits a disconnect event before switching chain.
MetaMask/metamask-extension#13375 (comment)
Until this gets fixed, I'm adding a shim to help prevent this from happening. #190

**Comment by tmm** [comment]
Yup, that's correct. It looks like it only occurs when you use chains not built-in to MetaMask.

**Comment by tmm** [comment]
Looks like this was an issue on MM's side and works again when testing (using this [CodeSandbox](https://uv07sn.csb.app/)).

@layinka can you try out and re-open this issue if it doesn't work?

**Comment by Pczek** [comment]
@tmm I still have this issue when switching between BSC and BSC Testnet. Not sure how related this is to "non default chains". 
Seems like I get as you already mentioned disconnected during the switching process.

This Error bubbels up, not sure where it is coming from though.

```
Error: underlying network changed (event="changed", network={"name":"bnb","chainId":56,"ensAddress":null,"_defaultProvider":null}, detectedNetwork={"name":"bnbt","chainId":97,"ensAddress":null,"_defaultProvider":null}, code=NETWORK_ERROR, version=providers/5.6.1)
    at Logger.makeError (index.js?dd68:219:1)
    at Web3Provider.eval (base-provider.js?4ba1:971:1)
    at Generator.next (<anonymous>)
    at fulfilled (base-provider.js?4ba1:5:43)
```

**Comment by richikchanda1999** [comment]
While adding connectors, there was an option named `shimChainChangedDisconnect` that had solved this issue. However, it was removed in 0.3.x and onwards.
This issue has resurfaced in the above sandbox as well.

> Looks like this was an issue on MM's side and works again when testing (using this [CodeSandbox](https://uv07sn.csb.app/)).
> 
> @layinka can you try out and re-open this issue if it doesn't work?

**Comment by tmm** [comment]
Yeah, the `shimChainChangedDisconnect` option was a hack. Better that this is fixed by MM. Will think about it some more.

**Comment by FranRom** [comment]
I'm facing the same issue in my project, any workaround meanwhile MM fix it?
Thanks in advance!

**Comment by richikchanda1999** [comment]
> I'm facing the same issue in my project, any workaround meanwhile MM fix it? Thanks in advance!

A workaround I used was to use a boolean state variable, with initial value as `false`. 
Its value changes from `false` to `true` when the `connect` function is called.  (`connect` from `useConnect` hook)
And from `true` to `false` when the `disconnect` function is called. (`disconnect` from `useDisconnect` hook)

The `useConnect` hook also returns a boolean value `isDisconnected`. 
So, a network change makes the value of `isDisconnected` false, however the state variable still has a value true.

**Comment by onigiri-x** [comment]
Hi there

I am having this issue on a webapp that has wagmi. Any options on how to fix this on the app - like a workaround?

I use this list of networks below, it  at first lets me connect to any of them. Then when I start like a transaction, it will randomnly change to whatever is the FIRST in the list, in this case bsc. Then I need to disconnect and reconnect to get it to work.

Any tips would be appreciated, thank you

const { chains, provider } = configureChains(
  [
      bsc, goerli, polygon, mainnet
  ],
  [
    publicProvider(),
  ]
);

const { connectors } = getDefaultWallets({
  appName: 'MyApp',
  chains
});

const wagmiClient = createClient({
  autoConnect: true,
  connectors,
  provider
});

**Comment by aryan877** [comment]
any fixes yet ?

**Comment by tmm** [comment]
Yup, was fixed in [0.12.6](https://github.com/wagmi-dev/wagmi/releases/tag/wagmi%400.12.6) by https://github.com/wagmi-dev/references/pull/177
